lsp-ido.el

```diff
@@ -0,0 +1,134 @@
+   "Prop"	   ; Property - 7
+   "Fld "	   ; Field - 8
+   "Cons"          ; Constructor - 9
+   "Enum"	   ; Enum - 10
+   "Intf"	   ; Interface - 11
+   "Func"          ; Function - 12
+   "Var "          ; Variable - 13
+   "Cnst"	   ; Constant - 14
+   "Str "	   ; String - 15
+   "Num "	   ; Number - 16
+   "Bool "	   ; Boolean - 17
+   "Arr "	   ; Array - 18
+   "Obj "	   ; Object - 19
+   "Key "	   ; Key - 20
+   "Null"	   ; Null - 21
+   "EmMm"	   ; EnumMember - 22
+   "Srct"	   ; Struct - 23
+   "Evnt"	   ; Event - 24
+   "Op  "          ; Operator - 25
+   "TPar"]          ; TypeParameter - 26
+  "A vector of 26 itens representing the SymbolKind."
+  :group 'lsp-ido
+  :type 'vector)
+
+(defcustom lsp-ido-show-symbol-filename
+  t
+  "Whether to show the project-relative path to a symbol's point of definition."
+  :group 'lsp-ido
+  :type 'boolean)
+
+(defcustom lsp-ido-show-symbol-kind
+  t
+  "Whether to show the symbol's kind when showing lsp symbols."
+  :group 'lsp-ido
+  :type 'boolean)
+
+(eval-when-compile
+  (lsp-interface
+   (lsp-ido:FormattedSymbolInformation
+    (:kind :name :location :textualRepresentation)
+    (:containerName :deprecated))))
+
+(lsp-defun lsp-ido--transform-candidate
+  ((symbol-information &as &SymbolInformation :kind :location (&Location :uri))
+   lsp-ido--results project-root)
+  (let* ((sanitized-kind (if (< kind (length lsp-ido-symbol-kind-to-string)) kind 0))
+	 (type (elt lsp-ido-symbol-kind-to-string sanitized-kind))
+	 (typestr (if lsp-ido-show-symbol-kind
+		      (format "[%s] " type)
+		    ""))
+	 (pathstr (if lsp-ido-show-symbol-filename
+		      (propertize (format " . %s" (file-relative-name (lsp--uri-to-path uri) project-root))
+				  'face 'font-lock-comment-face)
+		    ""))
+	 (textual-representation
+	  (lsp-render-symbol-information symbol-information "."))
+	 (entry (concat typestr textual-representation pathstr)))
+    (puthash entry symbol-information lsp-ido--results)))
+
+(lsp-defun lsp-ido--jump-selected-candidate
+  ((&SymbolInformation
+    :location (&Location :uri :range (&Range :start (&Position :line :character)))))
+  "Jump to selected candidate."
+  (find-file (lsp--uri-to-path uri))
+  (goto-char (point-min))
+  (forward-line line)
+  (forward-char character))
+
+(defun lsp-ido--workspace-symbol (workspaces)
+  "Search against WORKSPACES with PROMPT and INITIAL-INPUT."
+  (let* ((lsp-ido--results (make-hash-table :test 'equal))
+	 (workspace-root (lsp-workspace-root))
+	 (raw-choices
+	  (with-lsp-workspaces workspaces
+	    (lsp-request
+	     "workspace/symbol"
+	     (lsp-make-workspace-symbol-params :query "")))))
+    (mapc (lambda (it)
+	    (lsp-ido--transform-candidate it lsp-ido--results workspace-root))
+	  raw-choices)
+    lsp-ido--results))
+
+;;;###autoload
+(defun lsp-ido-workspace-symbol ()
+  "`ido' for lsp workspace/symbol."
+  (interactive)
+  (let* ((hash-table-candidates (lsp-ido--workspace-symbol (lsp-workspaces)))
```

Comment:
There should probably be a function for only the current workspace, and this command should then become `lsp-ido-global-workspace-symbol`.

Reply:
Looks like `lsp-workspaces` is the command for only the current workspace if present, look [here](https://github.com/emacs-lsp/lsp-mode/blob/fcea313af2aa2a5da7c1f10cc7462f0104681547/lsp-mode.el#L1262) and a parallel on lsp-ivy [here](https://github.com/emacs-lsp/lsp-ivy/blob/c70ee8b54357c56d1b972393ee53e57a2e545fbb/lsp-ivy.el#L209)

Reply:
And to be fair, I tested the `global-workspacee-symbol` option some times on `ivy` and didn't find this so useful. Because of the sync nature of `ido` I got several timeouts with projects no so big. I would rather stay with the `current-workspace`option for now.